docs/advanced-recipes/atom-creators.mdx

```diff
@@ -356,3 +356,66 @@ const ListOfThings = () => {
 ```
 
 <CodeSandbox id="ugkzm0" />
+
+## atomWithDebounce
+
+> `atomWithDebounce` helps with creating an atom where state set should be debounced.
+
+This can be especially useful for text search inputs, where you would like to fire an action only after the user has stopped typing for a delay, instead of firing an action on every keystroke.
+
+```tsx
+export function atomWithDebounce<Value>(
+  initialValue: Value,
+  delayMilliseconds = 500,
+  shouldDebounceOnReset = false
+) {
+  let timeout: NodeJS.Timeout
+
+  // DO NOT EXPORT currentValueAtom as using this atom to set state can cause
+  // inconsistent state between currentValueAtom and debouncedValueAtom
+  const currentValueAtom = atom(initialValue)
+
+  const debouncedValueAtom = atom(
+    initialValue,
+    (get, set, update: SetStateAction<Value>) => {
+      clearTimeout(timeout)
+      const prevValue = get(currentValueAtom)
+      const nextValue =
+        typeof update === 'function'
+          ? (update as (prev: Value) => Value)(prevValue)
+          : update
+
+      set(currentValueAtom, nextValue)
+
+      if (!shouldDebounceOnReset && nextValue === initialValue) {
+        set(debouncedValueAtom, nextValue)
+        return
+      }
+
+      set(currentValueAtom, nextValue)
+
+      timeout = setTimeout(() => {
+        set(debouncedValueAtom, nextValue)
+      }, delayMilliseconds)
+    }
+  )
+
+  const isDebouncingAtom = atom(
+    (get) => get(currentValueAtom) !== get(debouncedValueAtom)
+  )
+
+  return {
+    readOnlyCurrentValueAtom: atom((get) => get(currentValueAtom)),
+    isDebouncingAtom,
+    debouncedValueAtom,
+  }
+}
+```
+
+### Example Usage
+
+The sandbox link below shows how we would use a derived atom to fetch state based on the value of `debouncedValue`.
+
+When typing a pokemon's name in `<SearchInput>`, we do not fire a fetch on every letter, but only after `delayMilliseconds` has passed since the last text input.
+
+[CodeSandbox](https://codesandbox.io/s/withered-tdd-cjrz2y?file=/src/App.tsx)
```

Comment:
I got an error...
<img width="527" alt="image" src="https://user-images.githubusercontent.com/490574/183772894-acaa9ff4-ea4c-4cb5-a2a9-68016519cd78.png">

Reply:
the api only works for valid pokemon names, did not catch errors as it was just a demo haha 😆 try `ditto` or `charmander`

Reply:
Is it too hard to fix? Would be nice even if it's just a demo. Or, it's even nice because it's a *demo*.

Reply:
Did a hacky fix in sandbox here: https://codesandbox.io/s/withered-tdd-cjrz2y?file=/src/atoms.ts

Ideally should be done with `ErrorBoundary` but codesandbox seems to not work with ErrorBoundary well so had a hacky solution just for demo.